sympy/sets/fancysets.py

```diff
@@ -483,6 +492,108 @@ def _intersect(self, other):
             else:
                 return
 
+    def _union(self, other):
+        """
+
+        This function should only be used internally.
+        See Set._union for docstring
+
+        Used within the :class:`Union` class
+
+        Returns
+        =======
+        self._union(other) returns a new, joined ImageSet if self knows how
+        to join itself with other, otherwise it returns ``None``.
+
+        Examples
+        ========
+
+        >>> from sympy import Lambda
+        >>> from sympy.sets.fancysets import ImageSet
+        >>> from sympy import pprint, Union
+        >>> from sympy.core import Dummy, pi, S, Symbol
+        >>> n = Dummy('n')
+        >>> img1 = ImageSet(Lambda(n, 2*n*pi + 3*pi/4), S.Integers)
+        >>> img2 = ImageSet(Lambda(n, 2*n*pi + 7*pi/4), S.Integers)
+        >>> pprint(img1.union(img2), use_unicode=False)
+                3*pi
+        {n*pi + ---- | n in Integers()}
+                 4
+
+        """
+
+        from sympy.polys import Poly
+        from sympy.simplify import simplify
+        from sympy.polys.polyerrors import PolynomialError
+
+        if other is S.EmptySet:
+            return self
+
+        def first_val(imgset):
+            base = imgset.base_set
+            if base == S.Reals:
+                return imgset.lamda(0)
+            else:
+                iterable = iter(base)
+                # If base is not iterbale then TypeError.
+                return imgset.lamda(next(iterable))
+
+        def _attempt_1(self_expr, other_expr, var_self, var_other, base):
+            var_term = self_expr.coeff(var_self)
+            if var_term == 0:
+                return None
+            if simplify(
+                    other_expr.subs(var_other, var_self) -
+                    self_expr) % var_term == var_term / 2:
+                # (a*pi + pi + expr1) - ( a*pi  + exp1) == abs(a/2)*pi
+                # can be => (a/2)*pi + expr1
+                other_expr = (var_term / 2) * var_self + first_val(self)
+                ans = ImageSet(Lambda(var_self, other_expr), base)
+            elif simplify(
+                    other_expr.subs(var_other, var_self) -
+                    self_expr) % var_term == 0:
+                # (a*pi + expr1) - ( a*pi  + 2*pi + exp1) == 0
+                # can be => a*pi + expr1
+                other_expr = var_term * var_self + first_val(self)
+                ans = ImageSet(Lambda(var_self, other_expr), base)
+            else:
+                ans = None
+            return ans
+
+        if isinstance(other, ImageSet):
+            try:
```

Comment:
Please use `try`-`except` precisely on the lines you need to use them, otherwise they make it harder to debug the code. (I guess I've mentioned this already somewhere)